python35_filter.cpp

```diff
@@ -136,50 +356,76 @@ vector<Reading *>* Python35Filter::getFilteredReadings(PyObject* filteredData)
+ * Log an error from the Python interpreter
  */
 void Python35Filter::logErrorMessage()
 {
-#ifdef PYTHON_CONSOLE_DEBUG
-	// Print full Python stacktrace 
-	PyErr_Print();
-#endif
-	//Get error message
-	PyObject *pType, *pValue, *pTraceback;
-	PyErr_Fetch(&pType, &pValue, &pTraceback);
-	PyErr_NormalizeException(&pType, &pValue, &pTraceback);
-
-	PyObject* str_exc_value = PyObject_Repr(pValue);
-	PyObject* pyExcValueStr = PyUnicode_AsEncodedString(str_exc_value, "utf-8", "Error ~");
-
-	// NOTE from :
-	// https://docs.python.org/2/c-api/exceptions.html
-	//
-	// The value and traceback object may be NULL
-	// even when the type object is not.	
-	const char* pErrorMessage = pValue ?
-				    PyBytes_AsString(pyExcValueStr) :
-				    "no error description.";
-
-	Logger::getLogger()->fatal("An error occured in the Python script: %s",
-			pErrorMessage);
-
-	// Reset error
-	PyErr_Clear();
-
-	// Remove references
-	Py_CLEAR(pType);
-	Py_CLEAR(pValue);
-	Py_CLEAR(pTraceback);
-	Py_CLEAR(str_exc_value);
-	Py_CLEAR(pyExcValueStr);
+PyObject *ptype, *pvalue, *ptraceback;
+
+	if (PyErr_Occurred())
+	{
+		PyErr_Fetch(&ptype, &pvalue, &ptraceback);
+		PyErr_NormalizeException(&ptype,&pvalue,&ptraceback);
+
+		char *msg, *file, *text;
+		int line, offset;
+
+		int res = PyArg_ParseTuple(pvalue,"s(siis)",&msg,&file,&line,&offset,&text);
+
+		PyObject *line_no = PyObject_GetAttrString(pvalue,"lineno");
+		PyObject *line_no_str = PyObject_Str(line_no);
+		PyObject *line_no_unicode = PyUnicode_AsEncodedString(line_no_str,"utf-8", "Error");
+		char *actual_line_no = PyBytes_AsString(line_no_unicode);  // Line number
+
+		PyObject *ptext = PyObject_GetAttrString(pvalue,"text");
+		PyObject *ptext_str = PyObject_Str(ptext);
+		PyObject *ptext_no_unicode = PyUnicode_AsEncodedString(ptext_str,"utf-8", "Error");
+		char *error_line = PyBytes_AsString(ptext_no_unicode);  // Line in error
+
+		// Remove the trailing newline from the string
+		char *newline = rindex(error_line,  '\n');
+		if (newline)
+		{
+			*newline = '\0';
+		}
+
+		// Not managed to find a way to get the actual error message from Python
+		// so use the string representation of the Error class and tidy it up, e.g.
+		// SyntaxError('invalid syntax', ('/tmp/scripts/test_addition_script_script.py', 9, 1, '}\n'))
+		PyObject *pstr = PyObject_Repr(pvalue);
+		PyObject *perr = PyUnicode_AsEncodedString(pstr, "utf-8", "Error");
+		char *err_msg = PyBytes_AsString(perr);
+		char *end = index(err_msg, ',');
+		if (end)
+		{
+			*end = '\0';
+		}
+		end = index(err_msg, '(');
+		if (end)
+		{
+			*end = ' ';
+		}
+
+
+		if (error_line == NULL || actual_line_no == NULL || strcmp(error_line, "<NULL>") == 0
+			       	|| strcmp(actual_line_no, "<NULL>") == 0)
+		{
+			m_logger->error("Python error: %s in supplied script", err_msg);
+		}
+		else
+		{
+			m_logger->error("Python error: %s in %s at line %s of supplied script", err_msg, error_line, actual_line_no);
+		}
+
+		PyErr_Clear();
+	}
```

Comment:
I think we would need the Py_CLEAR statements for the various python objects used here.

Reply:
Well spotted